xresolver/xresolver.go

```diff
@@ -0,0 +1,111 @@
+package xresolver
+
+import (
+	"context"
+	"errors"
+	"net"
+	"strconv"
+	"sync"
+)
+
+// Note to self: Dial is not being set for net.Resolver because that is the Dial to the DNS server.
+
+var DefaultDialer = &net.Dialer{}
+
+type resolver struct {
+	resolvers map[Lookup]bool
+	lock      sync.RWMutex
+	dialer    *net.Dialer
```

Comment:
This works better as a non-pointer member.  Below, in `NewResolver`, you copy a pointer from outside the function into this function.  The `resolver` type thus holds a reference to an object outside its scope, which can lead to odd race conditions.

If this is a non-pointer member, the struct initializer in `NewResolver` will copy it completely, thus making a separate instance.  This makes for much more bulletproof code.